Big restructure

rjmassey opened the pull request:
Fixes express bugs, adds CIL mode, starts adding comments.

rjmassey commented:
Jacob, please can you OK this merger? I don't own the repo, so I can't. Thanks.

jkeger commented:
Hi, sorry I don't check my uni email early weekend mornings so I'm only seeing this now. Happy to merge if it's all working, though merging doesn't "do" anything so there's also no rush if there's anything to finish tweaking.

I've skimmed through some of the commits but it looks like there are quite a lot of changes, can you summarise what's happened? Any big revelations on our earlier issues? 

And also no rush but have you added documentation at least at the level of something quick in the functions where you've added/changed main arguments?

rjmassey commented:
Hi Jacob,

 If the cafe sci was yesterday, how did it go?


I’m just keen to reconcile the branches before you make a lot more changes, and they diverge significantly. Sorry, I only get nights and odd sessions to work on this, so apologies for all the stray emails at odd hours. I’d be very keen to skype during the week and pass you back the baton.


I have made changes to the express matrix, to allow different options. For example, trap occupancy can be remembered between iterations of express. The trailing of trails now works. Also can look at postag stamp cutouts froma larger image or small ROIs within a large image. And express correctly doesn’t over-empty the image. I also added something from pre-C++ code, which treats the first pixel to pixel transfer differently than the rest. The comments discuss it (I agree - we need a LOT more comments in the code).
I *think* the blooming now works. It had been grabbing electrons from anywhere in a pixel, rather than only those near the top. 
Which of the unit tests compare output against that from the C++ code? I haven’t yet found any.

In general, I’m finding so much code duplication to be a real problem, and would like to prioritise streamlining. You suggested waiting until later to fix it, but I’m not sure how much progress we can make without the code being logically simpler. For example:
a) the TrapNonUniformHeightDistribution and TrapManagerNonUniformHeightDistribution classes contain hundreds of lines but only one line is added to vanilla Traps. 
b) electrons_captured_in_pixel loops twice over all the watermarks (once in electrons_captured_by_traps) then again in updated_watermarks_from_capture/_not_enough). Effectively the same calculation of which traps can get an electron is done twice, but with different code.
c) updated_watermarks_from_capture and updated_watermarks_from_capture_not_enough are almost duplicates of each other (and both are duplicated again in the track_time class)

I’d like to change some variable/function names the to the kind of terminology electrical engineers use. That way people will stand a better chance of understanding it in the future. You mentioned search and replaces, but I tried a couple and its pretty difficult because of all the similar/extended variable names. Have you an easy method in mind?


Ideas/questions/tasks:

We can't get away without the Slow Capture traps recovering the behaviour of InstantCapture Traps, in the limit that their tau_capture is set to 0. I’d rather change the behaviour of slow traps, but if that is impossible (surely not) and we have to change the fast traps, can you propose an algorithm for fast capture that would reproduce the slow capture with tau_c=0?


At the end of electrons_captured_by_traps, rather than returning a number, then later having to go back and recalculate which watermarks to change, how about updating the watermark structure (to add the new watermak), and returning a numpy.ndarray which is a proposed delta to the occupancy levels in the watermark. From this, it would be easy to calculate the total number of electrons needed to implement the delta - and if there are not enough, to multiply the delta by some fraction before simply adding it to the watermarks. It will need a new routine to remove watermarks with occupancy=1, but both  updated_watermarks_from_capture and updated_watermarks_from_capture_not_enough can be removed. if we also change the watermark levels to be in units of number of exposed traps rather than fractional height within the CCD well, that would make it even easierl. But that’s just me.

How about changing the base Trap type to SlowCapture? This can handle the hard cases. Copy TrapInstantCapture (i.e. “Trap”) from that, and replace/reuse its routines with tau=0. This will remove code, and also make class inheritance possible for more complicated trap types that want e.g. both continuum and slowcapture.


Can we get rid of the nonuniform height classes? I reimplemented what is needed of them with a surface=True flag in the base Trap class, although I’m not sure how much one extra if statement slows things down. I’m also dithering about moving this to the trap manager rather than ccd_volume - or possibly even the trap. Calculating the fractional number of exposed surface traps needed knowledge of the trap density, so it is now done in the trap manager rather than (or as well as) ccd_volume. It’s a bit ugly, but it may be the only class where all the required information is available.

Consider whether we should move the watermarks from the trap managers into the traps? The watermarking scheme is only efficient if the watermarks are often reset to full occupancy. However, the densities of trap species can vary by a factor of 100. Does this mean the watermarks of low-density species are infrequently overwritten because traps in the high-dsnsity species are not yet full? (This will only happn in the not_enough regime). Maybe not a good idea.


I get what the OU people were saying about phases when we went there. it’s not sufficient to implement phases by cloning pixel,s if we are going to be able to implement trap pumping. Because trap pumping involves charge packets changing direction, we need to keep track of charge occupancy levels in multiple pixels. That could be easy to do, via a list of trap_managers. Rather than splitting the image into 4s, we split the watermarks into 4s. We can do nice things like say in timestep 1, electrons occupy phases 1&2, then in timestep 2, they occupy just phase 2, etc. For each tilmestep, we compute a new well filling function, and compare against one or more watermark arrays. Crucially, if we store enough phases (from about two pixels), we can shuffle charge back and forth through them


Longer term

It’s pretty slow with a reasonable sized image. I suspect it will benefit a lot from optimisation/cython. But not until the logical structure is simpler. 



Which 


> On 25 Apr 2020, at 12:12, Jacob Kegerreis <notifications@github.com> wrote:
> 
> 
> Hi, sorry I don't check my uni email early weekend mornings so I'm only seeing this now. Happy to merge if it's all working, though merging doesn't "do" anything so there's also no rush if there's anything to finish tweaking.
> 
> I've skimmed through some of the commits but it looks like there are quite a lot of changes, can you summarise what's happened? Any big revelations on our earlier issues?
> 
> And also no rush but have you added documentation at least at the level of something quick in the functions where you've added/changed main arguments?
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>
